### ISSUE TITLE
AMP tag doesn't render correctly in a sandbox-ed iframe

### ISSUE THREAD

**Issue by congyun** [ampproject/amphtml#2952]
AMP tag doesn't render in a sandbox-ed iframe.

## Description

For security issue, we need put amp tagged snippet into a sandbox-ed iframe. We've included the amp runtime js: "https://cdn.ampproject.org/v0.js" as a script into the iframe content, yet the amp tags, for example, amp-img cannot render in the sandbox-ed iframe except we put "allow-same-origin" into the sandbox attribute.

## How do we reproduce the issue?
As there are 2 files needed, jsbin cannot show it correctly...I attached the two files here. To reproduce, open "amptestParent.html" and click "Go". The amp-img is not rendered as it's in an iframe. 

[AMPTest.zip](https://github.com/ampproject/amphtml/files/228810/AMPTest.zip)

## What browsers are affected?

All browsers.

## Which AMP version is affected?

I think this is an existing issue for the latest amp version. I'm using the runtime JS directly from CDN.

**Comment by congyun** [comment]
FYI this is needed in a Google product, DoubleClick for Publisher. This is part of AMP for Ads effort and would be great if we can resolve asap as we are targeting a Q2 launch.

**Comment by rudygalfi** [comment]
I tried to reproduce and saw:

<img width="319" alt="screen shot 2016-04-21 at 9 12 04 am" src="https://cloud.githubusercontent.com/assets/3888212/14716295/3b481608-07a1-11e6-9363-a0cea64f0f2f.png">

It looks to me like the amp-img is loading. Am I missing the issue?

**Comment by congyun** [comment]
Thanks Rudy for the quick reply! Can I know your browser? I'm trying it on a latest Chrome but it doesn't work...neither on my Mac or workstation.

![unset 2](https://cloud.githubusercontent.com/assets/490849/14719218/09aadd42-07c9-11e6-9d68-2fb61d9b2a13.png)

![screen shot 2016-04-21 at 2 00 11 pm](https://cloud.githubusercontent.com/assets/490849/14719291/6d21b878-07c9-11e6-8a8c-ae6ef5f5627c.png)

Is there any other reason for AMP doesn't render?

**Comment by erwinmombay** [comment]
@congyun did you try adding "allow-same-origin" ? you usually need that for the origin to be passed during postMessage operations to the receiving frame when an iframe is sandboxed

**Comment by rudygalfi** [comment]
@congyun I tried this on Chrome 50.0.2661.75 desktop on Mac.

**Comment by congyun** [comment]
@erwinmombay Hi Erwin, it works for adding "allow-same-origin", yet for security concerns we are not supposed to put "allow-same-origin" in our sandboxed iframe per discussion with security review team, as the content in iframe is based on user input. So we are asking if AMP tag can be rendered in sandbox-ed iframe without "allow-same-origin".

**Comment by congyun** [comment]
@rudygalfi Thanks Rudy, I've tried again and asked my teammates to try it on, and we found that it works only after clearing browser data, or using an ignition window. My browser versions are: 50.0.2661.75 beta (64-bit) and 50.0.2661.86(64-bit). Is this expected? Could there be some cache issues with the runtime js?

Telling from the first error message in developer console, I'm not sure if amp-tag is officially supported in a sandbox-ed iframe without "allow-same-origin".

Thank you!

**Comment by rudygalfi** [comment]
I'm also on the AMP dev channel: https://cdn.ampproject.org/experiments.html

Can you try that too?

**Comment by congyun** [comment]
@rudygalfi Thank you! AMP dev channel works for me as well. Looks like it will always work after clearing browser data. I'm trying to see if it can work on our site.

**Comment by cramforce** [comment]
If the cookie stuff is the only problem, we should just wrap out cookie read/write operation in try-catch. Super simple change.

**Comment by congyun** [comment]
Thanks Malte and all! 

I tried inside our product and it works fine after clearing browser data as well. I'd consider this to be a corner case that doesn't block us. Would love to see that if there is change Malte mentioned above that can be applied to AMP, and prevent this issue comes again.

**Comment by congyun** [comment]
Hello AMP folks,

I'm seeing the issue in #2952 again in our DFP FE with AMP ads preview support. I'm wondering is there any recent change made here for the sandboxed-iframe?

The same reproduce procedure is still applied. I saw error in dev console as below:
![screen shot 2016-10-20 at 11 51 51 am](https://cloud.githubusercontent.com/assets/490849/19567555/1ef17058-96bc-11e6-9979-3fa4dcfda43c.png)
 
I also commented on the https://github.com/ampproject/amphtml/pull/3005 feel free to merge.

Our product (AMP for Native V2 Ads) is in beta, so it would be great if you can help, thanks!

**Comment by cramforce** [comment]
Closing in favor of other issue.

**Comment by cramforce** [comment]
@zhouyx The `this.isLocalStorageSupported_ = 'localStorage' in this.win;` line in `storage-impl.` needs to be wrapped in a try block. We should probably add an integration test for this.

**Comment by jridgewell** [comment]
> The this.isLocalStorageSupported_ = 'localStorage' in this.win; line in storage-impl. needs to be wrapped in a try block

Damn, I left that but forgot to submit the review. 😢

**Comment by cramforce** [comment]
I actually accidentally fixed this in #5602 :)

**Comment by zhouyx** [comment]
LOL 😄

**Comment by jridgewell** [comment]
@cramforce: How? It's the same `'localStorage' in this.win`?

**Comment by cramforce** [comment]
The code used to actually read the property as in `!!this.win.localStorage`. I changed it, because reading the `localStorage` is as slow as actually reading a value in Chrome.
